src/sentry/auth/helper.py

```diff
@@ -551,104 +504,64 @@ class AuthHelper:
+    def get_provider(self, provider_key):
+        if self.provider_model:
+            return self.provider_model.get_provider()
         elif provider_key:
-            provider = manager.get(provider_key)
+            return super().get_provider(provider_key)
         else:
             raise NotImplementedError
 
-        self.provider = provider
-        if flow == self.FLOW_LOGIN:
-            self.pipeline = provider.get_auth_pipeline()
-        elif flow == self.FLOW_SETUP_PROVIDER:
-            self.pipeline = provider.get_setup_pipeline()
+    def get_pipeline_views(self):
+        if self.flow == self.FLOW_LOGIN:
+            return self.provider.get_auth_pipeline()
+        elif self.flow == self.FLOW_SETUP_PROVIDER:
+            return self.provider.get_setup_pipeline()
         else:
             raise NotImplementedError
 
-        # we serialize the pipeline to be [AuthView().get_ident(), ...] which
-        # allows us to determine if the pipeline has changed during the auth
-        # flow or if the user is somehow circumventing a chunk of it
-        self.signature = md5_text(" ".join(av.get_ident() for av in self.pipeline)).hexdigest()
+    def is_valid(self):
+        return super().is_valid() and self.state.flow in (self.FLOW_LOGIN, self.FLOW_SETUP_PROVIDER)
 
-    def pipeline_is_valid(self):
-        if not self.state.is_valid():
-            return False
-        if self.state.flow not in (self.FLOW_LOGIN, self.FLOW_SETUP_PROVIDER):
-            return False
-        return self.state.signature == self.signature
-
-    def init_pipeline(self):
-        self.state.regenerate(
-            {
-                "uid": self.request.user.id if self.request.user.is_authenticated else None,
-                "auth_provider": self.auth_provider.id if self.auth_provider else None,
-                "provider_key": self.provider.key,
-                "org_id": self.organization.id,
-                "step_index": 0,
-                "signature": self.signature,
-                "flow": self.flow,
-                "data": {},
-            }
-        )
+    def get_initial_state(self):
+        state = super().get_initial_state()
+        state.update({"flow": self.flow})
+        return state
 
     def get_redirect_url(self):
         return absolute_uri(reverse("sentry-auth-sso"))
 
-    def clear_session(self):
-        self.state.clear()
-
-    def current_step(self):
-        """
-        Render the current step.
-        """
-        step_index = self.state.step_index
-
-        if step_index == len(self.pipeline):
-            return self.finish_pipeline()
-
-        return self.pipeline[step_index].dispatch(request=self.request, helper=self)
-
-    def next_step(self):
-        """
-        Render the next step.
-        """
-        self.state.step_index += 1
-        return self.current_step()
+    def dispatch_to(self, step: View):
+        return step.dispatch(request=self.request, helper=self)
```

Comment:
This is a weird little kludge. The overridden method calls the second kwarg `pipeline` instead of `helper`. AFIACT, that's a hard-coded parameter name in a lot of existing view classes. An overridden `dispatch_to` method seemed preferable to changing them all.